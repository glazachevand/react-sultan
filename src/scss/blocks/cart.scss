@import "./../base/vars";

.cart {
  &__container {
    width: 100%;
    padding-bottom: 100px;
    @media (max-width: $xl) {
      padding-bottom: 50px;
    }
  }
  &__title {
    margin-bottom: 50px;
    @media (max-width: $xl) {
      margin-bottom: 20px;
    }
  }
  &__product-row {
    display: flex;
    align-items: center;
    column-gap: 5%;
    padding: 50px 0;
    border-top: 1px dashed rgba(63, 78, 101, 0.3);
    border-bottom: 1px dashed rgba(63, 78, 101, 0.3);
    @media (max-width: $xl) {
      flex-direction: column;
      row-gap: 15px;
      padding: 25px 0;
    }
    @media (max-width: $md) {
      padding: 20px 0;
    }
  }
  &__actions {
    flex: 0 0 36.7%;
    display: flex;
    justify-content: space-between;
    align-items: center;
    column-gap: 43px;
    @media (max-width: $xxl) {
      width: 60%;
    }
    @media (max-width: $md) {
      column-gap: 15px;
      width: 100%;
    }
  }
  &__count-line {
    width: 1px;
    height: 59px;
    border-left: 1px dashed rgba(63, 78, 101, 0.3);
    @media (max-width: $xl) {
      &:first-child {
        display: none;
      }
    }
  }
  &__counts {
    flex: 0 0 auto;
    display: inline-flex;
    justify-content: space-between;
    width: 115px;
    line-height: 23px;
  }
  &__price-item {
    font-size: 1.875rem;
    font-weight: bold;
    line-height: 59px;
    white-space: nowrap;
    text-align: center;
    color: $dark;
    @media (max-width: $xxl) {
      font-size: 1.5rem;
    }
    @media (max-width: $md) {
      font-size: 1rem;
    }
  }
  &__remove-btn {
    flex: 0 0 auto;
  }
  &__bottom {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 50px;
    @media (max-width: $md) {
      flex-direction: column;
      row-gap: 25px;
      margin-top: 25px;
    }
  }
  &__order {
    padding: 21px 37.5px;
    font-weight: 700;
    @media (max-width: $md) {
      padding: 31.5px 86.5px;
    }
  }
  &__price-total {
    font-size: 1.875rem;
    font-weight: 700;
    line-height: 1.3;
    color: $dark;
    @media (max-width: $md) {
      order: -1;
    }
  }
}
.cart-product {
  display: flex;
  align-items: center;
  column-gap: 20px;
  @media (max-width: $md) {
    flex-direction: column;
    row-gap: 10px;
  }

  &__img {
    flex: 0 0 281px;
    height: 191px;
    @media (max-width: $md) {
      flex: 0 0 197px;
      width: 290px;
    }
    img {
      width: 100%;
      height: 100%;
      object-fit: contain;
    }
  }
  &__content {
    flex: 1 1 auto;
  }
  &__size {
    margin-bottom: 9px;
    font-size: 12px;
    line-height: 1.33;
    &::before {
      display: inline-block;
      margin-right: 8px;
      font-family: "icons";
      opacity: 0.23;
    }
    &._type1::before {
      content: "\e90f";
      font-size: 10px;
    }
    &._type2::before {
      content: "\e910";
      font-size: 20px;
    }
  }
  &__title {
    display: block;
    margin-bottom: 10px;
    font-size: 1.875rem;
    font-weight: 700;
    line-height: 1.2;
    color: $dark;
    @media (max-width: $xxl) {
      font-size: 1.5rem;
    }
    @media (max-width: $md) {
      font-size: 1.25rem;
      font-weight: 500;
    }
  }
  &__description {
    font-size: 12px;
    font-weight: 300;
    line-height: 1.5;
  }
}
