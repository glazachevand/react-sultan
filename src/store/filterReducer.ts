import { FilterState, FilterAction, FiltersActionTypes } from '../types/filter';
import { getManufactures } from '../utils/getManufactures';
import data from './../data/data.json';

const initialState: FilterState = {
  sort: 'price_asc',
  category: 'Все',
  priceMin: 10,
  priceMax: 10000,
  manufacturers: getManufactures(data.products),
  page: 1,
  filterProducts: data.products,
};

export const filterReducer = (state = initialState, action: FilterAction): FilterState => {
  let productsCopy;
  switch (action.type) {
    case FiltersActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };

    case FiltersActionTypes.SET_SORT:
      productsCopy = [...state.filterProducts];

      switch (action.payload) {
        case 'title_asc':
          productsCopy = productsCopy.sort((a, b) => `${a.brand} ${a.title}`.localeCompare(`${b.brand} ${b.title}`));
          break;

        case 'title_desc':
          productsCopy = productsCopy.sort((a, b) => `${b.brand} ${b.title}`.localeCompare(`${a.brand} ${a.title}`));
          break;

        case 'price_asc':
          productsCopy = productsCopy.sort((a, b) => a.price - b.price);
          break;

        case 'price_desc':
          productsCopy = productsCopy.sort((a, b) => b.price - a.price);
          break;

        default:
          break;
      }

      return { ...state, sort: action.payload, filterProducts: productsCopy };

    case FiltersActionTypes.SET_PAGE:
      return { ...state, page: action.payload };

    case FiltersActionTypes.FILTER_PRODUCTS:
      productsCopy = [...action.payload.products];

      if (action.payload.category !== 'Все') {
        productsCopy = productsCopy.filter((product) => product.typecare.includes(action.payload.category));
      }

      productsCopy = productsCopy.filter((product) => product.price < action.payload.priceMax && product.price > action.payload.priceMin);

      productsCopy = productsCopy.filter((product) => action.payload.manufacturers.includes(product.manufacturer));

      return { ...state, filterProducts: productsCopy };

    case FiltersActionTypes.SET_PARAMETERS:
      return { ...state, priceMin: action.payload.priceMin, priceMax: action.payload.priceMax, manufacturers: action.payload.manufacturers };

    case FiltersActionTypes.CLEAR_PARAMETERS:
      return { ...state, priceMin: 10, priceMax: 10000, manufacturers: action.payload };

    default:
      return state;
  }
};
