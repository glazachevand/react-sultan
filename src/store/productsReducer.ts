import { ProductsState, ProductsAction, ProductsActionTypes } from '../types/products';
import { getProductsFromLS } from './../utils/getProductsFromLS';

const initialState: ProductsState = getProductsFromLS();

export const productsReducer = (state = initialState, action: ProductsAction): ProductsState => {
  let arrCopy, index;
  switch (action.type) {
    case ProductsActionTypes.DELETE_PRODUCT:
      return { ...state, products: [...state.products.filter((product) => product.id !== action.payload.id)] };

    case ProductsActionTypes.ADD_PRODUCT:
      return { ...state, products: [...state.products, action.payload] };

    case ProductsActionTypes.UPDATE_PRODUCT:
      arrCopy = [...state.products];
      index = arrCopy.findIndex((item) => item.id === action.payload.id);
      arrCopy.splice(index, 1, action.payload);
      return { ...state, products: arrCopy };

    case ProductsActionTypes.DELETE_CATEGORY:
      arrCopy = [...state.categories].filter((item) => item !== action.payload);
      return { ...state, categories: arrCopy };

    case ProductsActionTypes.UPDATE_CATEGORIES:
      return { ...state, categories: action.payload };

    case ProductsActionTypes.ADD_CATEGORY:
      return { ...state, categories: [...state.categories, action.payload] };

    default:
      return state;
  }
};
