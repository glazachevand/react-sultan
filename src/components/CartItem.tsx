import React from 'react';
import { Link } from 'react-router-dom';
import CountContainer from './UI/CountContainer/CountContainer';
import RemoveBtn from './UI/RemoveBtn/RemoveBtn';
import { useActions } from '../hooks/useActions';
import { CartItemType } from '../types/cart';

const CartItem: React.FC<CartItemType> = (cartItem) => {
  const productImg = require(`./../pages/img/products/${cartItem.url}`);
  const { deleteCartItem } = useActions();

  const onClickDelete = () => {
    deleteCartItem(cartItem);
  };

  return (
    <div className="cart__product-row">
      <div className="cart-product">
        <Link to={`/product/${cartItem.id}`} className="cart-product__img" ><img src={productImg} alt="product" /></Link >
        <div className="cart-product__content">
          <div className={cartItem.typesize === 'вес' ? `cart-product__size _type2` : `cart-product__size _type1`}>{cartItem.size}</div>
          <Link to={`/product/${cartItem.id}`} className="cart-product__title"><h2><span>{cartItem.brand}</span> {cartItem.title}</h2></Link >
          <p className="cart-product__description">{cartItem.description}</p>
        </div>
      </div>
      <div className="cart__actions">
        <div className="cart__count-line"></div>
        <CountContainer id={cartItem.id} />
        <div className="cart__count-line"></div>
        <div className="cart__price-item">{cartItem.price * cartItem.count} ₽</div>
        <div className="cart__count-line"></div>
        <RemoveBtn action={onClickDelete} cssClass={'cart__remove-btn'} />
      </div>
    </div>
  );
};

export default CartItem;