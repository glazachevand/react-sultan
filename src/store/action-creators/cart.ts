import { CartsAction, CartActionTypes } from './../../types/cart';
import { CartItemType } from './../../types/cart';

export function addCartItem(item: CartItemType): CartsAction {
  return { type: CartActionTypes.ADD_ITEM, payload: item };
}

export function minusCartItem(item: CartItemType): CartsAction {
  return { type: CartActionTypes.MINUS_ITEM, payload: item };
}

export function deleteCartItem(item: CartItemType): CartsAction {
  return { type: CartActionTypes.DELETE_ITEM, payload: item };
}

export function clearCartItems(): CartsAction {
  return { type: CartActionTypes.CLEAR_ITEMS };
}
