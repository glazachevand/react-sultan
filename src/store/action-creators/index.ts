import * as ProductsActionCreators from './products';
import * as FilterActionCreators from './filter';
import * as CartActionCreators from './cart';

export default {
  ...ProductsActionCreators,
  ...FilterActionCreators,
  ...CartActionCreators,
};
