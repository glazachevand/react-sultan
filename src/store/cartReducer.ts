import { CartState, CartsAction, CartActionTypes } from '../types/cart';
import { calcTotalPrice } from '../utils/calcTotalPrice';
import { getCartFromLS } from '../utils/getCartFromLS';

const initialState: CartState = getCartFromLS();

export const cartReducer = (state = initialState, action: CartsAction): CartState => {
  let totalPrice, itemsCopy, findItem;

  switch (action.type) {
    case CartActionTypes.ADD_ITEM:
      itemsCopy = [...state.items];
      findItem = itemsCopy.find((item) => item.id === action.payload.id);

      if (findItem) {
        findItem.count++;
      } else {
        itemsCopy.push({ ...action.payload, count: 1 });
      }

      totalPrice = calcTotalPrice(itemsCopy);
      return { ...state, items: itemsCopy, totalPrice };

    case CartActionTypes.DELETE_ITEM:
      const itemsFilter = [...state.items.filter((item) => item.id !== action.payload.id)];
      totalPrice = calcTotalPrice(itemsFilter);
      return { ...state, items: itemsFilter, totalPrice };

    case CartActionTypes.MINUS_ITEM:
      itemsCopy = [...state.items];
      findItem = itemsCopy.find((item) => item.id === action.payload.id);

      if (findItem) {
        findItem.count--;
      }

      totalPrice = calcTotalPrice(itemsCopy);
      return { ...state, items: itemsCopy, totalPrice };

    case CartActionTypes.CLEAR_ITEMS:
      return { ...state, items: [], totalPrice: 0 };

    default:
      return state;
  }
};
