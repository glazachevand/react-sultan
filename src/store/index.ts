import { createStore, combineReducers } from 'redux';
import { productsReducer } from './productsReducer';
import { filterReducer } from './filterReducer';
import { composeWithDevTools } from '@redux-devtools/extension';
import { cartReducer } from './cartReducer';

export const rootReducer = combineReducers({
  productsRed: productsReducer,
  filter: filterReducer,
  cart: cartReducer,
});

export const store = createStore(rootReducer, composeWithDevTools());
export type RootState = ReturnType<typeof rootReducer>;
