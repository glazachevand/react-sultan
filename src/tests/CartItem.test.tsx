import { render, screen, fireEvent } from '@testing-library/react'
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router-dom';
import { store } from '../store';
import CartItem from '../components/CartItem';
import { CartItemType } from '../types/cart';

describe('Test cart item', () => {
  let cartItem: CartItemType;

  beforeEach(() => {
    cartItem = {
      "id": 2,
      "title": "Шампунь beauty-full volume плотность и объем",
      "url": "product2.webp",
      "brand": "Tresemme",
      "description": "Головокружительный объем от самых корней?",
      "typesize": "объем",
      "size": "650 мл",
      "price": 400,
      "count": 10
    }
  });

  test('Click on delete button', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/cart']}>
          <CartItem {...cartItem} />
        </MemoryRouter>
      </Provider>
    );
    const deleteBtn = screen.getByTestId('delete-btn');
    expect(deleteBtn).toBeInTheDocument();
    expect(screen.getByText(/Шампунь beauty-full/i)).toBeInTheDocument();
    fireEvent.click(deleteBtn);
    expect(screen.queryByTestId(/Шампунь beauty-full/i)).toBeNull();
  });

  test('Сheck totalprice', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/cart']}>
          <CartItem {...cartItem} />
        </MemoryRouter>
      </Provider>
    );

    const totalPrice = screen.getByTestId('totalPrice');
    expect(totalPrice).toHaveTextContent('4000');
    expect(totalPrice).toMatchSnapshot();
  });
});