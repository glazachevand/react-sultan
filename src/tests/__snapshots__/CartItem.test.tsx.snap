// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cart item Сheck totalprice 1`] = `
<div
  class="cart__price-item"
  data-testid="totalPrice"
>
  4000
   ₽
</div>
`;
